cli/doctest_runner.rs

```diff
@@ -0,0 +1,439 @@
+use regex::Regex;
+use std::collections::HashSet;
+use std::ffi::OsStr;
+use std::path::PathBuf;
+
+use crate::fs as deno_fs;
+use crate::installer::is_remote_url;
+use crate::test_runner::is_supported;
+
+pub struct DocTest {
+  // This removes repetition of imports in a file
+  imports: HashSet<String>,
+  // This contains codes in an @example section with their imports removed
+  bodies: Vec<DocTestBody>,
+}
+
+struct DocTestBody {
+  caption: String,
+  line_number: usize,
+  path: String,
+  value: String,
+  ignore: bool,
+  is_async: bool,
+}
+
+pub fn prepare_doctests(
+  mut include: Vec<String>,
+  root_path: &PathBuf,
+) -> Vec<DocTest> {
+  include.retain(|n| !is_remote_url(n));
+
+  let mut prepared = vec![];
+
+  for path in include {
+    let p = deno_fs::normalize_path(&root_path.join(path));
+    if p.is_dir() {
+      let test_files = deno_fs::files_in_subtree(p, |p| {
+        let valid_ext = ["ts", "tsx", "js", "jsx"];
+        p.extension()
+          .and_then(OsStr::to_str)
+          .map(|ext| valid_ext.contains(&ext) && !is_supported(p))
+          .unwrap_or(false)
+      });
+      prepared.extend(test_files);
+    } else {
+      prepared.push(p);
+    }
+  }
+
+  prepared
+    .iter()
+    .filter_map(|dir| {
+      // TODO(iykekings) use deno error instead
+      let content = std::fs::read_to_string(&dir)
+        .unwrap_or_else(|_| panic!("File doesn't exist {}", dir.display()));
+      extract_jsdoc_examples(content, dir.to_owned())
+    })
+    .collect::<Vec<_>>()
+}
+
+fn extract_jsdoc_examples(input: String, p: PathBuf) -> Option<DocTest> {
+  lazy_static! {
+    static ref JS_DOC_PATTERN: Regex =
+      Regex::new(r"/\*\*\s*\n([^\*]|\*[^/])*\*/").unwrap();
+    // IMPORT_PATTERN doesn't match dynamic imports by design
+    static ref IMPORT_PATTERN: Regex =
+      Regex::new(r"import[^(].*\n").unwrap();
+    static ref EXAMPLE_PATTERN: Regex = Regex::new(r"@example\s*(?:<\w+>.*</\w+>)*\n(?:\s*\*\s*\n*)*```").unwrap();
```

Comment:
It'd be great if you could annotate each regex with docstring showing examples of matching text - it's hard to grep it at first glance